File: components/users/src/com/github/joerodriguez/sbng2ex/invitation/InvitationService.java
package com.github.joerodriguez.sbng2ex.invitation;

import com.github.joerodriguez.sbng2ex.User;
import com.github.joerodriguez.sbng2ex.UserService;
import com.github.joerodriguez.sbng2ex.service.ServiceResponse;
import com.github.joerodriguez.sbng2ex.transaction.Transactions;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class InvitationService {

    private final Transactions transactions;
    private final PasswordGenerator passwordGenerator;
    private final EmailService emailService;
    private final UserService userService;

    @Autowired
    public InvitationService(
            Transactions transactions,
            PasswordGenerator passwordGenerator,
            EmailService emailService,
            UserService userService
    ) {
        this.transactions = transactions;
        this.passwordGenerator = passwordGenerator;
        this.emailService = emailService;
        this.userService = userService;
    }

    public ServiceResponse<User> invite(InvitationRequest invitationRequest) {
        String password = passwordGenerator.get();

        return transactions.create(t -> {
            t.addWithEntity(userService.create(invitationRequest.getEmail(), password));
            t.add(emailService.sendInvitation(invitationRequest.getEmail(), password));
        });
    }

}

File: components/users/src/com/github/joerodriguez/sbng2ex/invitation/InvitationService.kt
package com.github.joerodriguez.sbng2ex.invitation

import com.github.joerodriguez.sbng2ex.User
import com.github.joerodriguez.sbng2ex.UserService
import com.github.joerodriguez.sbng2ex.service.ServiceResponse
import com.github.joerodriguez.sbng2ex.transaction.Transactions
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class InvitationService

    @Autowired
    constructor(
        private val transactions: Transactions,
        private val passwordGenerator: PasswordGenerator,
        private val emailService: EmailService,
        private val userService: UserService
    )

{

    fun invite(invitationRequest: InvitationRequest) = transactions.create<User> {
        val password = passwordGenerator.get()

        it.addWithEntity(userService.create(invitationRequest.email, password))
        it.add(emailService.sendInvitation(invitationRequest.email, password))
    }

}
